File: 21.재귀/4779.칸토어 집합/main.py
import sys


def removeCenter(start, end):
    global last
    length = end - start + 1
    if length == 1:
        print(" " * (start - last - 1), end="")
        print("-" * (end - start + 1), end="")
        last = end
        return

    cutLength = length // 3
    cl = start + cutLength
    cr = end - cutLength
    removeCenter(start, cl - 1)
    removeCenter(cr + 1, end)


if __name__ == "__main__":
    lines = sys.stdin.readlines()
    for line in lines:
        N = int(line)
        k = 3**N
        last = 0

        removeCenter(1, k)
        print()

File: 21.재귀/4779.칸토어 집합/Main.java
import java.io.*;

public class Main {
    static int last = 0;
    static StringBuilder sb = new StringBuilder();

    static void removeCenter(int start, int end) {
        int len = end - start + 1;
        if (len == 1) {
            // 전꺼 출력
            for (int i = last + 1; i < end; i++) {
                sb.append(" ");
            }
            // 이번꺼 출력
            for (int i = start; i <= end; i++) {
                sb.append("-");
                last = end;
            }
            return;
        }

        int centerLen = len / 3;
        int cStart = start + centerLen, cEnd = end - centerLen;
        removeCenter(start, cStart - 1);
        removeCenter(cEnd + 1, end);

    }

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        String line;
        while (((line = br.readLine()) != null)) {
            int N = Integer.parseInt(line);
            int start = 1, end = (int) Math.pow(3, N);
            removeCenter(start, end);
            sb.append("\n");
        }

        bw.write(sb.toString());
        bw.close();
        br.close();
    }
}